src/stirling/source_connectors/socket_tracer/protocols/mqtt/stitcher_test.cc

```diff
@@ -0,0 +1,623 @@
+
+  int t = 0;
+  connect_frame.timestamp_ns = ++t;
+  pingreq_frame.timestamp_ns = ++t;
+  pingresp_frame.timestamp_ns = ++t;
+  connack_frame.timestamp_ns = ++t;
+  req_map[0].push_back(connect_frame);
+  resp_map[0].push_back(connack_frame);
+  req_map[0].push_back(pingreq_frame);
+  resp_map[0].push_back(pingresp_frame);
+  result = StitchFrames(&req_map, &resp_map);
+  EXPECT_TRUE(AreAllDequesEmpty(resp_map));
+  EXPECT_EQ(TotalDequeSize(req_map), 0);
+  EXPECT_EQ(result.error_count, 0);
+  EXPECT_EQ(result.records.size(), 2);
+
+  // Delayed subscribe response
+  t = 0;
+  sub_frame.timestamp_ns = ++t;
+  pub2_frame.timestamp_ns = ++t;
+  pubrec_frame.timestamp_ns = ++t;
+  pubrel_frame.timestamp_ns = ++t;
+  suback_frame.timestamp_ns = ++t;
+  pubcomp_frame.timestamp_ns = ++t;
+  unsub_frame.timestamp_ns = ++t;
+  unsuback_frame.timestamp_ns = ++t;
+  req_map[1].push_back(sub_frame);
+  req_map[1].push_back(pub2_frame);
+  req_map[1].push_back(pubrel_frame);
+  req_map[2].push_back(unsub_frame);
+  resp_map[1].push_back(pubrec_frame);
+  resp_map[1].push_back(suback_frame);
+  resp_map[1].push_back(pubcomp_frame);
+  resp_map[2].push_back(unsuback_frame);
+  result = StitchFrames(&req_map, &resp_map);
+  EXPECT_TRUE(AreAllDequesEmpty(resp_map));
+  EXPECT_EQ(TotalDequeSize(req_map), 0);
+  EXPECT_EQ(result.error_count, 0);
+  EXPECT_EQ(result.records.size(), 4);
+  req_map.clear();
+  resp_map.clear();
+
+  // Delayed responses with interleaved requests
+  t = 0;
+  pub1_frame.timestamp_ns = ++t;
+  pub2_pid2_frame.timestamp_ns = ++t;
+  pub1_pid3_frame.timestamp_ns = ++t;
+  puback_frame.timestamp_ns = ++t;
+  pubrec_pid2_frame.timestamp_ns = ++t;
+  pubrel_pid2_frame.timestamp_ns = ++t;
+  pubcomp_pid2_frame.timestamp_ns = ++t;
+  puback_pid3_frame.timestamp_ns = ++t;
+  req_map[1].push_back(pub1_frame);
+  req_map[2].push_back(pub2_pid2_frame);
+  req_map[3].push_back(pub1_pid3_frame);
+  req_map[2].push_back(pubrel_pid2_frame);
+  resp_map[1].push_back(puback_frame);
+  resp_map[2].push_back(pubrec_pid2_frame);
+  resp_map[2].push_back(pubcomp_pid2_frame);
+  resp_map[3].push_back(puback_pid3_frame);
+  result = StitchFrames(&req_map, &resp_map);
+  EXPECT_TRUE(AreAllDequesEmpty(resp_map));
+  EXPECT_EQ(TotalDequeSize(req_map), 0);
+  EXPECT_EQ(result.error_count, 0);
+  EXPECT_EQ(result.records.size(), 4);
+  req_map.clear();
+  resp_map.clear();
+}
+
+TEST(MqttStitcherTest, DummyResponseStitching) {
+  // Test for requests that do not have responses
+  absl::flat_hash_map<packet_id_t, std::deque<Message>> req_map;
+  absl::flat_hash_map<packet_id_t, std::deque<Message>> resp_map;
+
+  RecordsWithErrorCount<Record> result;
+
+  Message pub0_frame, disconnect_frame, auth_frame, connect_frame, connack_frame;
+  pub0_frame = CreateFrame(kRequest, 3, 0, 0);         // PUBLISH with QoS 0
+  disconnect_frame = CreateFrame(kRequest, 14, 0, 0);  // DISCONNECT
+  auth_frame = CreateFrame(kRequest, 15, 0, 0);        // AUTH
```

Comment:
This actually complicates things. This would be similar to the case of the request-response being in the opposite direction, which would put the first AUTH packet sent by the server in the response deque, and the AUTH response by the client in the request deque. I have the logic of checking for packets with timestamps greater than the request frame in the response deque, which the server sent AUTH request wouldn't have.

Reply:
How important do you think it is to capture these AUTH frames? From my perspective, I think supporting these frames isn't the most critical for the initial support. I'd be ok with consuming / dropping them and leaving a TODO comment for now with a brief explanation of the challenges.

Reply:
Yeah I think we can skip doing stitching for Auth frames for now. The only time it is sent by the server is when authentication is not complete in the connection stage itself (I think). We can support them later.